src/charm.py

```diff
@@ -0,0 +1,217 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Charm the service."""
+
+import logging
+
+import ops
+from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus
+from ops.pebble import ExecError
+
+from literals import APPLICATION_PORT
+from relations.provider import LDAPProvider
+from state import State
+from utils import log_event_handler, random_string
+
+# Log messages can be retrieved using juju debug-log
+logger = logging.getLogger(__name__)
+
+
+class OpenLDAPK8SCharm(ops.CharmBase):
+    """Charm the service."""
+
+    def __init__(self, *args):
+        """Construct.
+
+        Args:
+            args: Ignore.
+        """
+        super().__init__(*args)
+        self._state = State(self.app, lambda: self.model.get_relation("peer"))
+        self.name = "openldap"
+
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(
+            self.on.openldap_pebble_ready, self._on_openldap_pebble_ready
+        )
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.restart_action, self._on_restart)
+        self.framework.observe(
+            self.on.get_admin_password_action, self._on_get_admin_password
+        )
+        self.framework.observe(
+            self.on.load_test_users_action, self._on_load_test_users
+        )
+        self.provider = LDAPProvider(self)
+
+    @log_event_handler(logger)
+    def _on_install(self, event):
+        """Install application.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        self.unit.status = MaintenanceStatus("installing OpenLdap")
+
+    @log_event_handler(logger)
+    def _on_openldap_pebble_ready(self, event: ops.PebbleReadyEvent):
+        """Define and start openldap using the Pebble API.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        self.update(event)
+
+    @log_event_handler(logger)
+    def _on_config_changed(self, event: ops.ConfigChangedEvent):
+        """Handle configuration changes.
+
+        Args:
+            event: The event triggered when the relation changed.
+        """
+        self.update(event)
+
+    @log_event_handler(logger)
+    def _on_load_test_users(self, event):
```

Comment:
maybe i misunderstood this... the test users file is already present in the charm from the beginning and this action just loads it? I don't see the need of an action if that is done once and the users can't be modified..couldn't this action take a diff input and flush it?

Reply:
No you didn't misunderstand, this action is purely for ease of testing in ranger integration tests without having to install `ldap-utils` and provide a file. The recommended method for making any updates to the ldap is via `ldapadd`as I don't think the LDAP user management should be handled by the charm. The charm should manage the application but I think the application should be used to manage the users. 

I've added an option to provide an alternative file to the one pre loaded for testing purposes. But since this charm is a minimal one for dev and staging testing until we get something from IS I don't know that it makes sense to further develop the functionality. At least not within the scope of this work.

Reply:
i guess you can merge it now, but i see this charm more like an mvp then a dev thing... you can sell it to IS for some nice comments in the performance review :)

Reply:
Yeah I guess it is not far off a DOG1. haha, you are welcome IS. Though I think in a staging or production environment I would probably remove this action entirely, as I don't think it makes sense to be managing users via the charm when the application is providing an interface for managing ldap users.